README.md
# LTHRadioButton

Slightly inspired by Google's material radio button.

The clip below has 3 sections: full speed, 25% and 10%, but after converting it to GIF, it actually made it longer, so the 10% part takes a really long time. [Here's](https://rolandleth.com/images/radio-button/video.mp4) an mp4 link; try with Chrome if Safari doesn't work - for me it doesn't.

![](https://rolandleth.com/images/radio-button/gif.gif)

## How to install

#### CocoaPods

[CocoaPods](https://cocoapods.org) is a dependency manager for Cocoa projects. You can install it with the following terminal command:

```
gem install cocoapods
```

To integrate `LTHRadioButton` into your Xcode project using CocoaPods, specify it in your Podfile:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'
use_frameworks!

target '<Your Target Name>' do
  pod 'LTHRadioButton'
end
```

Then, run the following terminal command:

```
pod install
```

#### Manual

Drag `LTHRadioButton.swift` from the `source` folder into your Xcode project.

## How to use

The initializer takes up to 3 params: a `diameter`, a `selectedColor`, and a `deselectedColor`. All of them are optional:

* `diameter` defaults to `18`
* `selectedColor` defaults to a light blue
* `deselectedColor` defaults to `UIColor.lightGray`

It doesn't use Auto Layout internally, but after initialization it will have a proper size, so you can simply create constraints based on its `frame.width` and `frame.height`.

#### Properties

`selectedColor` and `deselectedColor` have been made publicly customizable for cases like a `tableView` with alternating row and radio colors, where the `tableView` might dequeue a cell with one color for displaying a cell with a different color.

`isSelected` - Indicates whether the radio button is selected.

`useTapGestureRecognizer` - Indicates whether a tap gesture recognizer should be added to the control when setting callbacks. This defaults to `true` just so that `onSelect` and `onDeselect` can add the gesture recognizer automatically, but the recognizer is **not** added by default.
  - Settings this to `true` will also add the required `UITapGestureRecognizer` if needed.
  - Settings this to `false` will also remove the `UITapGestureRecognizer` if it was previously added.

#### Methods

```swift
init(diameter: CGFloat = 18, selectedColor: UIColor? = nil, deselectedColor: UIColor? = nil) // Colors default internally if nil.
func select(animated: Bool = true) // Selects the radio button.
func deselect(animated: Bool = true) // Deselects the radio button.
```

### Callbacks

You can make use of the `onSelect` and `onDeselect` methods to add closures to be run when selecting/deselecting the control. Since these closures make most sense for taps and because there are no recognizers by default, these methods will also add one (and only one) `UITapGestureRecognizer` to the control to handle the taps; the closure calls happen right as the animations begin.

If you'd like to use the callbacks but don't need the tap gesture recognizer, you can set `useTapGestureRecognizer` to `false`.

### Example

```swift
let radioButton = LTHRadioButton(selectedColor: .red)
container.addSubview(radioButton)

radioButton.translatesAutoresizingMaskIntoConstraints = false
NSLayoutConstraint.activate([
  radioButton.centerYAnchor.constraint(equalTo: container.centerYAnchor),
  radioButton.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 16),
  radioButton.heightAnchor.constraint(equalToConstant: radioButton.frame.height),
  radioButton.widthAnchor.constraint(equalToConstant: radioButton.frame.width)]
)

radioButton.onSelect {
  print("I'm selected.")
}

radioButton.onDeselect {
  print("I'm deselected.")
}

[...]

radioButton.select() // I'm selected.

[...]

radioButton.deselect(animated: false) // I'm deselected.
```

## Apps using this control

If you're using this control, I'd love hearing from you!

## License
Licensed under MIT. If you'd like (or need) a license without attribution, don't hesitate to [contact me](mailto:roland@leth.ro).


CHANGELOG.md
# 1.1.2

* Added `onSelect` and `onDeselect` callbacks.
* Fixed `radius` to `diameter` parameter name.
* Better center positioning of subviews (https://github.com/rolandleth/LTHRadioButton/pull/5).

# 1.1.1

* Improved Objective-C naming.

# 1.1.0

* Interface builder suport added.
* Improved tests.

# 1.0.3

* Tests added.
* README updated.

# 1.0.2

* Pod fixed. Tag bumped to keep them in sync.

# 1.0.1

* README and project updates.

# 1.0.0

* Initial release.


source/LTHRadioButton.swift
//
//  LTHRadioButton.swift
//  LTHRadioButton
//
//  Created by Roland Leth on 17.11.2016.
//  Copyright © 2016 Roland Leth All rights reserved.
//

import UIKit

/// A radio button with a fill animation.
public class LTHRadioButton: UIView {
	
	/// The view used for the rippling effect.
	private let waveCircle   = UIView()
	/// The view used for the filling.
	private let circle       = UIView()
	/// The view used for the margin.
	private let innerCircle  = UIView()
	/// The color used for the selected state.
	public var selectedColor = UIColor(red: 0.29, green: 0.56, blue: 0.88, alpha: 1.0) {
		willSet {
			innerCircle.layer.borderColor = newValue.cgColor
			waveCircle.layer.borderColor  = newValue.cgColor
		}
	}
	/// The color used for the deselected state.
	public var deselectedColor = UIColor.lightGray {
		willSet {
			circle.layer.borderColor = newValue.cgColor
		}
	}
	/// A `Boolean` value that indicates whether the button is selected.
	public private(set) var isSelected = false
	/// A `Boolean` value that indicates whether the button should add a `UITapGestureRecognizer`.
	/// - Note: This defaults to `true` just so that `onSelect` and `onDeselect` can add the gesture recognizer automatically, but it is **not** added by default.
	/// - Settings this to `true` will also add the required `UITapGestureRecognizer` if needed.
	/// - Settings this to `false` will also remove the `UITapGestureRecognizer` if it was previously added.
	public var useTapGestureRecognizer = true {
		willSet {
			guard newValue else { return removeGestureRecognizer(tapGesture) }
			addTapGesture()
		}
	}
	
	/// The final width of the inner circle's border, used for filling.
	private var innerBorderWidth: CGFloat {
		return innerCircle.frame.width * 0.6
	}
	/// The percentage with which the innler circle will increase for the elastic filling effect.
	private let innerIncreaseDelta: CGFloat = 1.1
	/// The width of the inner circle after increasing for the elasting filling effect.
	private var innerIncreasedWidth: CGFloat {
		return innerCircle.frame.width * innerIncreaseDelta
	}
	
	/// The tap gesture that will handle the `onSelect`/`onDeselect` callbacks.
	private lazy var tapGesture: UITapGestureRecognizer = {
		let tg = UITapGestureRecognizer()
		
		tg.numberOfTapsRequired = 1
		tg.numberOfTouchesRequired = 1
		tg.addTarget(self, action: #selector(self.toggleState))
		
		return tg
	}()
	/// The closure that will be called when the control is selected.
	private var didSelect: () -> Void = { } {
		willSet {
			addTapGesture()
		}
	}
	/// The closure that will be called when the control is deselected.
	private var didDeselect: () -> Void = { } {
		willSet {
			addTapGesture()
		}
	}
	
	
	// MARK: - Callbacks
	
	/// Sets a closure that will be called when the control is selected.
	///
	/// - Important: Calling this will also add the required `UITapGestureRecognizer`, unless it was already added or `useTapGestureRecognizer` was set to `false`.
	/// - Parameter closure: The closure the be called.
	public func onSelect(execute closure: @escaping () -> Void) {
		didSelect = closure
	}
	
	/// Sets a closure that will be called when the control is deselected.
	///
	/// - Important: Calling this will also add the required `UITapGestureRecognizer`, unless it was already added or `useTapGestureRecognizer` was set to `false`.
	/// - Parameter closure: The closure the be called.
	public func onDeselect(execute closure: @escaping () -> Void) {
		didDeselect = closure
	}
	
	
	// MARK: - Select animations
	
	/// Initializes and returns the animation for filling the inner circle.
	private func innerBorderIncrease() -> CABasicAnimation {
		let borderWidth = CABasicAnimation(keyPath: "borderWidth")
		
		borderWidth.duration       = 0.2
		borderWidth.fromValue      = 0.0
		borderWidth.toValue        = innerBorderWidth
		borderWidth.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseIn)
		borderWidth.fillMode       = kCAFillModeBackwards
		borderWidth.beginTime      = CACurrentMediaTime()
		
		return borderWidth
	}
	
	/// Initializes and returns the animation group for increasing the inner circle,
	/// the first step in giving it an elastic effect.
	private func innerIncreaseGroup() -> CAAnimationGroup {
		let group = CAAnimationGroup()
		
		let bounds       = CABasicAnimation(keyPath: "bounds")
		bounds.fromValue = NSValue(cgRect: innerCircle.frame)
		bounds.toValue   = NSValue(cgRect: CGRect(
			x: 0, y: 0,
			width: innerIncreasedWidth, height: innerIncreasedWidth)
		)
		
		let cornerRadius       = CABasicAnimation(keyPath: "cornerRadius")
		cornerRadius.fromValue = innerCircle.layer.cornerRadius
		cornerRadius.toValue   = innerCircle.layer.cornerRadius * innerIncreaseDelta
		
		group.duration       = 0.1
		group.animations     = [bounds, cornerRadius]
		group.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
		group.beginTime      = CACurrentMediaTime() + 0.23
		
		return group
	}
	
	/// Initializes and returns the animation group for decreasing the inner circle,
	/// the second step in giving it an elastic effect.
	private func innerDecreaseGroup() -> CAAnimationGroup {
		let group = CAAnimationGroup()
		
		let bounds = CABasicAnimation(keyPath: "bounds")
		bounds.toValue   = NSValue(cgRect: innerCircle.frame)
		bounds.fromValue = NSValue(cgRect: CGRect(
			x: 0, y: 0,
			width: innerIncreasedWidth, height: innerIncreasedWidth)
		)
		
		let cornerRadius       = CABasicAnimation(keyPath: "cornerRadius")
		cornerRadius.fromValue = innerCircle.layer.cornerRadius * innerIncreaseDelta
		cornerRadius.toValue   = innerCircle.layer.cornerRadius
		
		group.duration       = 0.15
		group.animations     = [bounds, cornerRadius]
		group.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
		group.beginTime      = CACurrentMediaTime() + 0.31
		
		return group
	}
	
	/// Initializes and returns the animation for coloring the outer circle.
	private func circleBorderColor() -> CABasicAnimation {
		let borderColor = CABasicAnimation(keyPath: "borderColor")
		
		borderColor.duration       = 0.15
		borderColor.fromValue      = deselectedColor.cgColor
		borderColor.toValue        = selectedColor.cgColor
		borderColor.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionLinear)
		borderColor.fillMode       = kCAFillModeBackwards
		borderColor.beginTime      = CACurrentMediaTime() + 0.28
		
		return borderColor
	}
	
	/// Initializes and returns the animation group for increasing the wave circle,
	/// giving it a pulsing effect.
	private func waveIncreaseGroup() -> CAAnimationGroup {
		let start = 0.21
		let delta = CGFloat(2.15)
		let width = waveCircle.frame.width * delta
		let group = CAAnimationGroup()
		
		let bounds       = CABasicAnimation(keyPath: "bounds")
		bounds.fromValue = NSValue(cgRect: waveCircle.frame)
		bounds.toValue   = NSValue(cgRect: CGRect(
			x: 0, y: 0,
			width: width, height: width)
		)
		
		let cornerRadius       = CABasicAnimation(keyPath: "cornerRadius")
		cornerRadius.fromValue = waveCircle.layer.cornerRadius
		cornerRadius.toValue   = waveCircle.layer.cornerRadius * delta
		
		group.duration       = 0.25
		group.animations     = [bounds, cornerRadius]
		group.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
		group.beginTime      = CACurrentMediaTime() + start
		
		return group
	}
	
	/// Initializes and returns the animation for fading out the wave circle.
	private func waveAlphaDecrease() -> CABasicAnimation {
		let opacity = CABasicAnimation(keyPath: "opacity")
		
		opacity.duration       = 0.31
		opacity.fromValue      = 0.3
		opacity.toValue        = 0
		opacity.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
		opacity.beginTime      = CACurrentMediaTime() + 0.26
		
		return opacity
	}
	
	/// Initializes and returns the animation for decreasing width of the wave circle,
	/// the final step in giving it a ripple effect.
	private func waveBorderDecrease() -> CABasicAnimation {
		let borderWidth = CABasicAnimation(keyPath: "borderWidth")
		
		borderWidth.duration       = 0.26
		borderWidth.fromValue      = frame.width * 0.3
		borderWidth.toValue        = 0
		borderWidth.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
		borderWidth.beginTime      = CACurrentMediaTime() + 0.29
		
		return borderWidth
	}
	
	
	// MARK: - Deselect animations
	
	/// Initializes and returns the animation group for emptying the inner circle.
	private func innerDecreaseGroupReverse(duration: Double) -> CAAnimationGroup {
		let group = CAAnimationGroup()
		
		let borderWidth       = CABasicAnimation(keyPath: "borderWidth")
		borderWidth.fromValue = innerBorderWidth
		borderWidth.toValue   = 0.0
		
		let opacity       = CABasicAnimation(keyPath: "opacity")
		opacity.fromValue = 1.0
		opacity.toValue   = 0.0
		
		group.duration       = duration
		group.animations     = [borderWidth, opacity]
		group.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseIn)
		group.beginTime      = CACurrentMediaTime()
		
		return group
	}
	
	/// Initializes and returns the animation group for decoloring the outer circle.
	private func circleBorderColorReverse(duration: Double) -> CABasicAnimation {
		let borderColor = CABasicAnimation(keyPath: "borderColor")
		
		borderColor.duration       = duration
		borderColor.fromValue      = selectedColor.cgColor
		borderColor.toValue        = deselectedColor.cgColor
		borderColor.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
		borderColor.beginTime      = CACurrentMediaTime()
		
		return borderColor
	}
	
	
	// MARK: - Actions
	
	/// Sets the selected state of the control.
	///
	/// - Parameter animated: A `Boolean` value which determines whether the transition should be animated or not. Defaults to `true`.
	@objc(selectAnimated:)
	public func select(animated: Bool = true) {
		guard !isSelected else { return }
		isSelected = true
		
		innerCircle.layer.borderWidth = innerBorderWidth
		circle.layer.borderColor      = selectedColor.cgColor
		
		didSelect()
		
		guard animated else { return }
		
		innerCircle.layer.add(innerBorderIncrease(), forKey: "innerBorderWidth")
		innerCircle.layer.add(innerIncreaseGroup(), forKey: "innerIncreaseGroup")
		innerCircle.layer.add(innerDecreaseGroup(), forKey: "innerDecreaseGroup")
		
		circle.layer.add(circleBorderColor(), forKey: "circleBorderColor")
		
		waveCircle.layer.add(waveIncreaseGroup(), forKey: "innerDecreaseGroup")
		waveCircle.layer.add(waveAlphaDecrease(), forKey: "outerAlphaDecrease")
		waveCircle.layer.add(waveBorderDecrease(), forKey: "outerBorderDecrease")
	}
	
	/// Sets the deselected state of the control.
	///
	/// - Parameter animated: A `Boolean` value which determines whether the transition should be animated or not. Defaults to `true`.
	@objc(deselectAnimated:)
	public func deselect(animated: Bool = true) {
		guard isSelected else { return }
		isSelected = false
		
		removeAnimations()
		setDeselectedEndValues()
		didDeselect()
		
		guard animated else { return }
		
		let duration = 0.2
		innerCircle.layer.add(innerDecreaseGroupReverse(duration: duration), forKey: "innerDecreaseGroupReverse")
		circle.layer.add(circleBorderColorReverse(duration: duration), forKey: "circleBorderColorReverse")
	}
	
	/// Sets the end values for the deselected state.
	private func setDeselectedEndValues() {
		innerCircle.layer.borderWidth = 0
		circle.layer.borderColor      = deselectedColor.cgColor
	}
	
	/// Removes all animations.
	private func removeAnimations() {
		waveCircle.layer.removeAllAnimations()
		circle.layer.removeAllAnimations()
		innerCircle.layer.removeAllAnimations()
	}
	
	/// Toggles between selected and deselected states.
	@objc private func toggleState() {
		guard isSelected else { return select() }
		
		deselect()
	}
	
	/// Adds the `UITapGestureRecognizer`.
	private func addTapGesture() {
		guard useTapGestureRecognizer else { return }
		guard gestureRecognizers?.contains(tapGesture) != true else { return }
		
		addGestureRecognizer(tapGesture)
	}
	
	
	// MARK: - Init
	
	/// Initializes and returns a radio button with a diameter, a selected color and a deselected color.
	///
	/// - Parameters:
	///   - diameter: A constant, indicating the diameter. Optional, defaults to `18`.
	///   - selectedColor: The `UIColor` used for the selected state. Optional, defaults to a light blue.
	///   - deselectedColor: The `UIColor` used for the deselected state. Optional, defaults to `.lightGray`.
	public init(diameter: CGFloat = 18, selectedColor: UIColor? = nil, deselectedColor: UIColor? = nil) {
		let size = CGSize(width: diameter, height: diameter)
		super.init(frame: CGRect(origin: .zero, size: size))
		commonInit(diameter: diameter, selectedColor: selectedColor, deselectedColor: deselectedColor)
	}
	
	/// Called when a radio button is loaded from a xib. `selectedColor` and `deselectedColor` can be either added as *User Defined Runtime Attributes*, or set after initialization.
	required public init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		commonInit(diameter: frame.size.width, selectedColor: nil, deselectedColor: nil)
	}
	
	/// Performs the initialization of the radio button.
	///
	/// - Parameters:
	///   - diameter: A constant, indicating the diameter.
	///   - selectedColor: The `UIColor` used for the selected state, defaults to a light blue if `nil`.
	///   - deselectedColor: The `UIColor` used for the deselected state, defaults to `.lightGray` if `nil`.
	private func commonInit(diameter: CGFloat, selectedColor: UIColor?, deselectedColor: UIColor?) {
		let innerCircleDiameter = diameter / 1.6
		let radius = diameter * 0.5
		let center = CGPoint(x: radius, y: radius)
		
		backgroundColor      = .clear
		self.selectedColor   = selectedColor ?? self.selectedColor
		self.deselectedColor = deselectedColor ?? self.deselectedColor
		
		addSubview(circle)
		circle.backgroundColor    = .clear
		circle.layer.cornerRadius = radius
		circle.layer.borderColor  = self.deselectedColor.cgColor
		circle.layer.borderWidth  = diameter * 0.1
		circle.frame.size         = CGSize(width: diameter, height: diameter)
		circle.center             = center
		
		addSubview(innerCircle)
		innerCircle.backgroundColor    = .clear
		innerCircle.layer.cornerRadius = innerCircleDiameter * 0.5
		innerCircle.layer.borderColor  = self.selectedColor.cgColor
		innerCircle.layer.borderWidth  = 0
		innerCircle.frame.size         = CGSize(width: innerCircleDiameter, height: innerCircleDiameter)
		innerCircle.center             = center
		
		addSubview(waveCircle)
		waveCircle.backgroundColor    = .clear
		waveCircle.layer.cornerRadius = innerCircle.layer.cornerRadius
		waveCircle.layer.borderColor  = self.selectedColor.cgColor
		waveCircle.layer.borderWidth  = 0
		waveCircle.alpha              = 0
		waveCircle.frame.size         = innerCircle.frame.size
		waveCircle.center             = center
	}
	
}


RadioButtonDemo/RadioButtonDemo/RadioCell.swift
//
//  RadioCell.swift
//  RadioButtonDemo
//
//  Created by Roland Leth on 19.10.2016.
//  Copyright © 2016 Roland Leth. All rights reserved.
//

import UIKit

class RadioCell: UITableViewCell {
	
	private let selectedColor   = UIColor(red: 74/255, green: 144/255, blue: 226/255, alpha: 1.0)
	private let deselectedColor = UIColor.lightGray
	
	private lazy var radioButton: LTHRadioButton = {
		let r = LTHRadioButton()
		self.contentView.addSubview(r)
		
		r.translatesAutoresizingMaskIntoConstraints = false
		NSLayoutConstraint.activate([
			r.centerYAnchor.constraint(equalTo: self.contentView.centerYAnchor),
			r.leadingAnchor.constraint(equalTo: self.contentView.leadingAnchor, constant: 16),
			r.heightAnchor.constraint(equalToConstant: r.frame.height),
			r.widthAnchor.constraint(equalToConstant: r.frame.width)]
		)
		
		return r
	}()
	
	func update(with color: UIColor) {
		backgroundColor             = color
		radioButton.selectedColor   = color == .darkGray ? .white : selectedColor
		radioButton.deselectedColor = color == .darkGray ? .lightGray : deselectedColor
	}
	
	override func setSelected(_ selected: Bool, animated: Bool) {
		super.setSelected(selected, animated: animated)
		
		if selected {
			return radioButton.select(animated: animated)
		}
		
		radioButton.deselect(animated: animated)
	}
}


RadioButtonDemo/RadioButtonDemo/ViewController.swift
//
//  ViewController.swift
//  RadioButtonDemo
//
//  Created by Roland Leth on 19.10.2016.
//  Copyright © 2016 Roland Leth. All rights reserved.
//

import UIKit

class ViewController: UITableViewController {

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return 20
	}
	
	
	// MARK: - Cells

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = tableView.dequeueReusableCell(withIdentifier: "cell",
		                                         for: indexPath) as! RadioCell
		cell.selectionStyle = .none
		
		return cell
	}
	
	override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
		(cell as? RadioCell)?.update(with: indexPath.row % 4 == 0 ? .darkGray : .white)
	}
	
	
	// MARK: - Selection
	
	override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
		tableView.cellForRow(at: indexPath)?.setSelected(true, animated: true)
		return indexPath
	}
	
	override func tableView(_ tableView: UITableView, willDeselectRowAt indexPath: IndexPath) -> IndexPath? {
		tableView.cellForRow(at: indexPath)?.setSelected(false, animated: true)
		return indexPath
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		let top = UIApplication.shared.statusBarFrame.height
		tableView.contentInset = UIEdgeInsets(top: top, left: 0, bottom: 0, right: 0)
		tableView.allowsMultipleSelection = false
		tableView.register(RadioCell.self, forCellReuseIdentifier: "cell")
	}
}



RadioButtonDemo/RadioButtonDemoTests/RadioButtonDemoTests.swift
//
//  RadioButtonDemoTests.swift
//  RadioButtonDemoTests
//
//  Created by Roland Leth on 03.12.2016.
//  Copyright © 2016 Roland Leth. All rights reserved.
//

import XCTest
import RadioButtonDemo

class RadioButtonDemoTests: XCTestCase {
	
	private let selectedColor = UIColor(red: 0.29, green: 0.56, blue: 0.88, alpha: 1.0)
	private lazy var mirror: Mirror = Mirror(reflecting: self.radioButton!)
	private var radioButton: LTHRadioButton!
	
	
	// MARK: -
	
	func testInitWithDefaultValues() {
		radioButton = LTHRadioButton()
		
		XCTAssertNotNil(radioButton)
		XCTAssertEqual(radioButton.selectedColor, selectedColor)
		XCTAssertEqual(radioButton.deselectedColor, .lightGray)
		XCTAssertEqual(radioButton.frame.size, CGSize(width: 18, height: 18))
	}
	
	func testInitWithParameters() {
		let selectedColor = UIColor.red
		let deselectedColor = UIColor.blue
		let diameter: CGFloat = 20
		
		radioButton = LTHRadioButton(diameter: diameter,
		                             selectedColor: selectedColor,
		                             deselectedColor: deselectedColor)
		
		XCTAssertNotNil(radioButton)
		XCTAssertEqual(radioButton.selectedColor, selectedColor)
		XCTAssertEqual(radioButton.deselectedColor, deselectedColor)
		XCTAssertEqual(radioButton.frame.size, CGSize(width: diameter, height: diameter))
	}
	
	func testXibLoading() {
		radioButton = Bundle.main
			.loadNibNamed("RadioButton", owner: self, options: nil)?
			.first as? LTHRadioButton
		
		let diameter: CGFloat = 24
		
		XCTAssertNotNil(radioButton, "Failed to load xib.")
		XCTAssertEqual(radioButton.selectedColor, UIColor.blue)
		XCTAssertEqual(radioButton.deselectedColor, UIColor.red)
		XCTAssertEqual(radioButton.frame.size, CGSize(width: diameter, height: diameter))
		
		// These, in conjunction with `testCircles`, are enough to assert that `commonInit` was properly ran.
		mirror.children.forEach {
			switch $0.label {
			case "waveCircle"?:
				guard let waveCircle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(waveCircle.layer.borderColor, UIColor.blue.cgColor)
			case "innerCircle"?:
				guard let innerCircle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(innerCircle.layer.borderColor, UIColor.blue.cgColor)
			case "circle"?:
				guard let circle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(circle.layer.cornerRadius, diameter * 0.5)
				XCTAssertEqual(circle.layer.borderColor, UIColor.red.cgColor)
				XCTAssertEqual(circle.layer.borderWidth, diameter * 0.1)
			default: return
			}
		}
	}
	
	func testCircles() {
		radioButton = LTHRadioButton()
		
		let diameter: CGFloat = 18
		let innerDiameter: CGFloat = diameter / 1.6
		let innerSize = CGSize(width: innerDiameter, height: innerDiameter)
		let innerCornerRadius: CGFloat = innerDiameter * 0.5
		
		mirror.children.forEach {
			switch $0.label {
			case "innerIncreaseDelta"?:
				XCTAssertEqual($0.value as? CGFloat, 1.1)
			case "waveCircle"?:
				guard let waveCircle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(waveCircle.layer.cornerRadius, innerCornerRadius)
				XCTAssertEqual(waveCircle.layer.backgroundColor, UIColor.clear.cgColor)
				XCTAssertEqual(waveCircle.layer.borderColor, selectedColor.cgColor)
				XCTAssertEqual(waveCircle.layer.borderWidth, 0)
				XCTAssertEqual(waveCircle.alpha, 0)
				XCTAssertEqual(waveCircle.frame.size, innerSize)
				XCTAssertEqual(waveCircle.center, radioButton.center)
			case "innerCircle"?:
				guard let innerCircle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(innerCircle.layer.cornerRadius, innerCornerRadius)
				XCTAssertEqual(innerCircle.layer.backgroundColor, UIColor.clear.cgColor)
				XCTAssertEqual(innerCircle.layer.borderColor, selectedColor.cgColor)
				XCTAssertEqual(innerCircle.layer.borderWidth, 0)
				XCTAssertEqual(innerCircle.frame.size, innerSize)
				XCTAssertEqual(innerCircle.center, radioButton.center)
			case "circle"?:
				guard let circle = $0.value as? UIView else { return XCTFail() }
				XCTAssertEqual(circle.layer.cornerRadius, diameter * 0.5)
				XCTAssertEqual(circle.layer.backgroundColor, UIColor.clear.cgColor)
				XCTAssertEqual(circle.layer.borderColor, UIColor.lightGray.cgColor)
				XCTAssertEqual(circle.layer.borderWidth, diameter * 0.1)
				XCTAssertEqual(circle.center, radioButton.center)
			default: return
			}
		}
	}
	
	func testSelectedDeselectedStates() {
		radioButton = LTHRadioButton()
		
		radioButton.select(animated: false)
		testState(selected: true)
		
		radioButton.deselect(animated: false)
		testState(selected: false)
	}
	
	func testProperties() {
		radioButton = LTHRadioButton()
		
		let expected: Set<String> = [
			"waveCircle", "circle", "innerCircle",
			"selectedColor", "deselectedColor",
			"isSelected", "innerIncreaseDelta",
			"useTapGestureRecognizer",
			"didSelect", "didDeselect", "tapGesture.storage" // .storage for lazy vars.
			// Computed properties aren't visible in mirrors.
//			"innerBorderWidth", "innerIncreasedWidth"
		]
		
		let actual = Set(mirror.children.flatMap { $0.label })
		let difference = actual.symmetricDifference(expected)
		
		// ¯\-(ツ)-/¯ Not super-useful, I know.
		XCTAssertTrue(difference.isEmpty, "Properties have changed: \(difference)")
	}
	
	func testOnSelectCallback() {
		var called = false
		radioButton = LTHRadioButton()
		
		radioButton.onSelect {
			called = true
		}
		
		XCTAssertFalse(called)
		
		radioButton.select(animated: false)
		
		XCTAssertTrue(called)
	}
	
	func testOnDeselectCallback() {
		var called = false
		radioButton = LTHRadioButton()
		
		radioButton.onDeselect {
			called = true
		}
		
		radioButton.select(animated: false)
		
		XCTAssertFalse(called)
		
		radioButton.deselect(animated: false)
		
		XCTAssertTrue(called)
	}
	
	func testUseRecognizerTrueAddsRecognizer() {
		radioButton = LTHRadioButton()
		radioButton.useTapGestureRecognizer = true
		
		XCTAssertEqual(1, radioButton.gestureRecognizers?.count)
	}
	
	func testUseRecognizerFalseBeforeCallbacksDoesNotAddRecognizer() {
		radioButton = LTHRadioButton()
		radioButton.useTapGestureRecognizer = false
		
		radioButton.onSelect { }
		
		XCTAssertNil(radioButton.gestureRecognizers)
	}
	
	func testUseRecognizerFalseRemovesRecognizer() {
		radioButton = LTHRadioButton()
		
		radioButton.onSelect { }
		
		radioButton.useTapGestureRecognizer = false
		
		XCTAssertEqual(true, radioButton.gestureRecognizers?.isEmpty)
	}
	
	
	// MARK: - Helpers
	
	private func testState(selected: Bool, file: StaticString = #file, line: UInt = #line) {
		XCTAssertEqual(radioButton.isSelected, selected, file: file, line: line)
		
		mirror.children.forEach {
			switch $0.label {
			case "innerCircle"?:
				guard let innerCircle = $0.value as? UIView else {
					return XCTFail(file: file, line: line)
				}
				
				let expectedWidth: CGFloat = selected ? 6.75 : 0
				XCTAssertEqual(innerCircle.layer.borderWidth, expectedWidth,
				               "Wrong innerCircle border width.", file: file, line: line)
			case "circle"?:
				guard let circle = $0.value as? UIView else {
					return XCTFail(file: file,line: line)
				}
				
				let expectedColor: UIColor = selected ? selectedColor : .lightGray
				XCTAssertEqual(circle.layer.borderColor, expectedColor.cgColor,
				               "Wrong circle border color.", file: file, line: line)
			default: return
			}
		}
	}
	
}
